designdocs/user-handles.md

```diff
@@ -0,0 +1,323 @@
+The user handle system will also enforce guidelines to ensure that handles are user-friendly and easy to remember, as well as prevent misuse and abuse.
+
+The goal of this proposal is to enable users to create unique handles on the Frequency blockchain using a registry, making it more accessible for users to engage with the network.
+
+## Assumptions
+
+* ```MessageSourceAccount``` exists for a user to create a handle.
+* It is optional for a user to create a handle.
+* Handles are unique and has a 1:1 mapping with ```MessageSourceAccount```.
+* Governance will define the range of suffixes allowed for a given handle.
+* Governance will define the time period a handle must be retired before it can be reused.
+
+## Requirements
+
+The high level requirements for user handles are:
+
+* Support user ability to choose any valid handle and select a suffix from a set of options allowed by the chain.
+* Ensure handles and suffixes are unique and non-conflicting.
+* Make the system resistant to namespace exhaustion, homoglyphs and race conditions.
+* Make the system easy to use and integrate with existing UI and wallet systems.
+* Reduce the likelihood of abuse and misuse of the system for example secondary market trading of handles.
+
+## Proposal
+
+User handle ```registry``` on frequency chain.
+
+### Handle Guidelines
+
+* User handles must be unique and each msa can only be be mapped to one handle.
+* User handles must be between 2 and 32 characters (32b limit) long.
+* User handles cannot contain the following substrings (or homoglyphs): @, #, :, ., ```
+* User handles cannot be (@admin, @everyone, @all) *blocklist of handles we reserve
+* Suffixes are u32 limited to a range defined by governance
+* Suffix will be randomly constrained.
+* Homoglyph versions of handles should still resolve to the same MSA.
+* After a handle is released, it cannot be reused by a time block set by governance.
+
+### General Steps
+
+* Query the chain for the current seed value.
+* Hash the seed with the user desired handle base.
+* Apply the resulting value as the seed to the specified PRNG.
+* Generate the next 10/20 ```u32``` values from the PRNG these are the possible suffixes within the range of suffixes allowed by the chain.
+* Query the chain to check which values are available, check if the handle with the suffix is available.
+* Choose an available value and attempt to claim it by submitting the full handle with the suffix to the chain.
+* Chain validates the handle and suffix and maps the handle to the ```msa_id```.
+* Chain maintains a reverse mapping from handle to msa for offchain use.
+
+### Chain Steps
+
+* Every 100 blocks (e.g., when block_number % 100 == 1), update the current seed by taking the Merkle root of the previous block. Keep the current and one previous seed so that two seeds, s0 and s1, are part of the chain state at any given time.
```

Comment:
It sounds as if we need to calculate the Merkle root when it's actually just the hash of the previous block.  Correct?